# Итоги выполнения домашнего задания к уроку 1

Задача 1. Подберите скорость обучения (alpha) и количество итераций для градиентного спуска (до совпадения ответов с результатами МНК). Как влияют друг на друга эти два параметра?
Описаие исходной задачи:

In [1]:
import numpy as np

In [2]:
X = np.array([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1],   # для умножения на intercept
              [1, 1, 2, 1, 3, 0, 5, 10, 1, 2]]).T # стаж репетитора

In [3]:
# средний балл ЕГЭ (целевая переменная)
y = [45, 55, 50, 59, 65, 35, 75, 80, 50, 60]

In [4]:
def calc_mse(y, y_pred):
    err = np.mean((y - y_pred)**2)
    return err

In [5]:
n = X.shape[0]
alpha = 1e-3
w = np.array([1, 0.5])

for i in range(1000+1):
    y_pred = np.dot(w, X.T)
    err = calc_mse(y, y_pred)
    for j in range(w.shape[0]):
        w[j] -= alpha * (1/n * 2 * np.sum(X[:,j] * (y_pred - y)))
    if i % 100 == 0:
        print(i, w, err) # МНК: array([47.23214286,  3.91071429]) 45.937499999999986

0 [1.1102 0.84  ] 3173.15
100 [ 7.43515582 10.71754582] 899.5990444359252
200 [11.44993589 10.49282195] 735.5540619653672
300 [14.98921624  9.8634316 ] 606.0048772266812
400 [18.17511588  9.27626113] 500.7989288428639
500 [21.04606457  8.74617136] 415.355423398327
600 [23.63334499  8.26841272] 345.9620646005472
700 [25.96499229  7.83785632] 289.6039104863001
800 [28.06626432  7.44984037] 243.83236130353217
900 [29.95992301  7.10016181] 206.65877484533297
1000 [31.66648131  6.78503276] 176.46806566379882


На первом этапе меняла параметры количества итераций для получения результата (при каждой последуюшей итерации коэффициенты и mse меняются незначительно, результат близок к результату, полученному методом наименьших квадратов) без изменения скорости обучения. Пробовала 3000 итераций, 5000 итераций, 10000 итерций и 20000. Результат получен только при использовании 20000 итераций:

In [6]:
n = X.shape[0]
alpha = 1e-3
w = np.array([1, 0.5])

for i in range(20000+1):
    y_pred = np.dot(w, X.T)
    err = calc_mse(y, y_pred)
    for j in range(w.shape[0]):
        w[j] -= alpha * (1/n * 2 * np.sum(X[:,j] * (y_pred - y)))
    if i % 2000 == 0:
        print(i, w, err) # МНК: array([47.23214286,  3.91071429]) 45.937499999999986

0 [1.1102 0.84  ] 3173.15
2000 [41.73215361  4.92632937] 62.2342399245102
4000 [46.54546925  4.03751382] 46.19152576288303
6000 [47.14641166  3.92654521] 45.94145963171208
8000 [47.22143932  3.91269078] 45.937561720839334
10000 [47.23080652  3.91096105] 45.93750096207482
12000 [47.23197602  3.91074509] 45.937500014996345
14000 [47.23212203  3.91071813] 45.93750000023376
16000 [47.23214026  3.91071477] 45.93750000000364
18000 [47.23214253  3.91071435] 45.937500000000036
20000 [47.23214282  3.91071429] 45.937499999999986


При сохранении скорости обучения на первоначальном уровне для получения результата количество итераций желательно не менее 10000. Уже после 7800 итераций mse меняется незначительно,но близкий к искомому результат (47.23,  3.910) получается только после приблизительно 10000 итераций, а точный до практически последнего предложенного знака после запятой - после 20000 итераций.

Следующим этапом пробую при сохранении количества итераций изменить скорость обучения. Беру значения 0,01, 0,05 и 0,07:

Пример 1:

In [7]:
n = X.shape[0]
alpha = 1e-2
w = np.array([1, 0.5])

for i in range(1000+1):
    y_pred = np.dot(w, X.T)
    err = calc_mse(y, y_pred)
    for j in range(w.shape[0]):
        w[j] -= alpha * (1/n * 2 * np.sum(X[:,j] * (y_pred - y)))
    if i % 100 == 0:
        print(i, w, err) # МНК: array([47.23214286,  3.91071429]) 45.937499999999986

0 [2.102 3.9  ] 3173.15
100 [31.88770806  6.74418155] 175.19445858001853
200 [41.83683774  4.90699865] 61.9177717428135
300 [45.33508261  4.26102097] 47.913169919666785
400 [46.56511152  4.03388672] 46.181755648107604
500 [46.99760587  3.95402334] 45.96769776787538
600 [47.14967657  3.92594232] 45.941233404700036
700 [47.20314662  3.91606866] 45.93796156758051
800 [47.2219474   3.91259695] 45.93755706443538
900 [47.228558    3.91137626] 45.937507054979434
1000 [47.23088237  3.91094704] 45.937500872219864


Пример 2:

In [8]:
n = X.shape[0]
alpha = 5e-2
w = np.array([1, 0.5])

for i in range(1000+1):
    y_pred = np.dot(w, X.T)
    err = calc_mse(y, y_pred)
    for j in range(w.shape[0]):
        w[j] -= alpha * (1/n * 2 * np.sum(X[:,j] * (y_pred - y)))
    if i % 100 == 0:
        print(i, w, err) # МНК: array([47.23214286,  3.91071429]) 45.937499999999986

0 [ 6.51 17.5 ] 3173.15
100 [47.03142979  3.94777749] 45.96159894380851
200 [47.23117922  3.91089223] 45.93750055548498
300 [47.23213823  3.91071514] 45.93750000001278
400 [47.23214283  3.91071429] 45.93749999999999
500 [47.23214286  3.91071429] 45.937499999999986
600 [47.23214286  3.91071429] 45.937499999999986
700 [47.23214286  3.91071429] 45.9375
800 [47.23214286  3.91071429] 45.9375
900 [47.23214286  3.91071429] 45.9375
1000 [47.23214286  3.91071429] 45.9375


Пример 3:

In [9]:
n = X.shape[0]
alpha = 7e-2
w = np.array([1, 0.5])

for i in range(1000+1):
    y_pred = np.dot(w, X.T)
    err = calc_mse(y, y_pred)
    for j in range(w.shape[0]):
        w[j] -= alpha * (1/n * 2 * np.sum(X[:,j] * (y_pred - y)))
    if i % 100 == 0:
        print(i, w, err) # МНК: array([47.23214286,  3.91071429]) 45.937499999999986

0 [ 8.714 24.3  ] 3173.15
100 [ 90145.66877384 487925.52597244] 3006560937495.4287
200 [3.42406430e+09 1.85427696e+10] 4.3422928006532086e+21
300 [1.30126711e+14 7.04691687e+14] 6.271453384410737e+30
400 [4.94528129e+18 2.67808092e+19] 9.057686655059327e+39
500 [1.87938409e+23 1.01776671e+24] 1.3081766300802822e+49
600 [7.14233298e+27 3.86787820e+28] 1.8893633227335074e+58
700 [2.71434246e+32 1.46993231e+33] 2.7287551873415993e+67
800 [1.0315474e+37 5.5862695e+37] 3.941065639863664e+76
900 [3.92024973e+41 2.12298258e+42] 5.6919720940028355e+85
1000 [1.48983537e+46 8.06809452e+46] 8.220757855742908e+94


Судя по примеру 3,при существенном увеличении скорости обучения, алгоритм не работает.

Но судя по примеру 1 (результат на 1000 итерации) и примеру 2 (результат на 300 итерации) можно увеличить скорость обучения алгоритма, которая позволит снизить количество итераций. Т.о. чем больше скорость обучения, тем меньше итераций необходимо для получения результата, но слишком большие скорости обучения могут привести к тому, что алгоритм "проскочит" нужный результат.

Итоговый подбор параметров для предложенного примера алгоритма (изменение одновременно параметра скорости обучения (увелиение до 0,06) и количества итераций (уменьшение до 500)):

In [10]:
n = X.shape[0]
alpha = 6e-2
w = np.array([1, 0.5])

for i in range(500+1):
    y_pred = np.dot(w, X.T)
    err = calc_mse(y, y_pred)
    for j in range(w.shape[0]):
        w[j] -= alpha * (1/n * 2 * np.sum(X[:,j] * (y_pred - y)))
    if i % 50 == 0:
        print(i, w, err) # МНК: array([47.23214286,  3.91071429]) 45.937499999999986

0 [ 7.612 20.9  ] 3173.15
50 [45.5816168   4.21574742] 47.603575735386116
100 [47.16625286  3.92288138] 45.94015500316733
150 [47.22951256  3.91119999] 45.937504230928766
200 [47.23203786  3.91073367] 45.937500006742276
250 [47.23213867  3.91071506] 45.93750000001074
300 [47.23214269  3.91071432] 45.93750000000002
350 [47.23214285  3.91071429] 45.93749999999999
400 [47.23214286  3.91071429] 45.9375
450 [47.23214286  3.91071429] 45.937500000000014
500 [47.23214286  3.91071429] 45.93749999999999


Задачи 2. В этом коде мы избавляемся от итераций по весам, но тут есть ошибка, исправьте ее:

In [39]:
alpha = 1e-3
w = np.array([1, 0.5])
for i in range(20000+1):
    y_pred = np.dot(w, X.T)
    err = calc_mse(y, y_pred)
    '''for j in range(w.shape[0]):
        w[j] -= alpha * (1/n * 2 * np.sum(X[j] * (y_pred - y)))'''
     
    new_w = w - alpha * 1/n * 2 * np.dot(X.T, (np.dot(w, X.T) - y))# ошибка была в данной формуле, изменила ее, предлагаю на этом шаге находить новое значение вектора w
    w = new_w # добавила данную строчку, чтобы новый вектор w заходил на следующую итерацию
    if i % 2000 == 0:
        print(i, w, err) # [47.23214286  3.91071429] 45.937499999999986

0 [1.1102 0.84  ] 3173.15
2000 [41.73215361  4.92632937] 62.2342399245102
4000 [46.54546925  4.03751382] 46.19152576288303
6000 [47.14641166  3.92654521] 45.94145963171208
8000 [47.22143932  3.91269078] 45.937561720839334
10000 [47.23080652  3.91096105] 45.93750096207482
12000 [47.23197602  3.91074509] 45.937500014996345
14000 [47.23212203  3.91071813] 45.93750000023376
16000 [47.23214026  3.91071477] 45.93750000000364
18000 [47.23214253  3.91071435] 45.937500000000036
20000 [47.23214282  3.91071429] 45.937499999999986


Задача 3. Вместо того, чтобы задавать количество итераций, задайте условие остановки алгоритма - когда ошибка за итерацию начинает изменяться ниже определенного порога.
Сколько нужно сделать итераций, если установить допустимое отклонение mse в размере diff=10−6
, а значение 𝛼=10−2?

Предлагаю следующий код для реализации алгоритма остановки и определения количества итераций для получения итогового результата. Алгоритм выводит две последние операции до остановки. Получилось 810 итераций.

In [56]:
w = np.array([1, 0.5])
diff = 1e-6
alpha = 1e-2
i = 0 # счетчик количества пройденных итераций
och = 1 # переменная, определяющая разницу между mse на текущем и предыдущем шаге/ итерации
err_list = []
i_list = []
w_list = []
while och >= diff:
    y_pred = np.dot(w, X.T)
    err = calc_mse(y, y_pred)
    err_list.append(err)
    new_w = w - alpha * 1/n * 2 * np.dot(X.T, (np.dot(w, X.T) - y))
    w_list.append(new_w)
    w = new_w
    i = i+1
    i_list.append(i)
    if len(err_list)> 2:
        och = err_list[-2]-err_list[-1]
print(i_list[-2], w_list[-2], err_list[-2]) 
print(i_list[-1], w_list[-1], err_list[-1]) # [47.23214286  3.91071429] 45.937499999999986

809 [47.22276525  3.91244593] 45.93754827651849
810 [47.22286276  3.91242793] 45.937547277798465
